File: src/elevator_press_button/src/elevator_press_button.cpp
// C++ standard headers
#include <exception>
#include <string>

// Boost headers
#include <boost/shared_ptr.hpp>

// ROS headers
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <control_msgs/FollowJointTrajectoryAction.h>
#include <ros/topic.h>


// Our Action interface type for moving TIAGo's head, provided as a typedef for convenience
typedef actionlib::SimpleActionClient<control_msgs::FollowJointTrajectoryAction> arm_control_client;
typedef boost::shared_ptr< arm_control_client>  arm_control_client_Ptr;


// Create a ROS action client to move TIAGo's arm
void createArmClient(arm_control_client_Ptr& actionClient)
{
  ROS_INFO("Creating action client to arm controller ...");

  actionClient.reset( new arm_control_client("/arm_controller/follow_joint_trajectory") );

  int iterations = 0, max_iterations = 3;
  // Wait for arm controller action server to come up
  while( !actionClient->waitForServer(ros::Duration(2.0)) && ros::ok() && iterations < max_iterations )
  {
    ROS_INFO("Waiting for the arm_controller_action server to come up");
    ++iterations;
  }

  if ( iterations == max_iterations )
    throw std::runtime_error("Error in createArmClient: arm controller action server not available");
}


// Generates a simple trajectory with two waypoints to move TIAGo's arm 
void waypoints_arm_goal(control_msgs::FollowJointTrajectoryGoal& goal)
{
  // The joint names, which apply to all waypoints
  goal.trajectory.joint_names.push_back("arm_1_joint");
  goal.trajectory.joint_names.push_back("arm_2_joint");
  goal.trajectory.joint_names.push_back("arm_3_joint");
  goal.trajectory.joint_names.push_back("arm_4_joint");
  goal.trajectory.joint_names.push_back("arm_5_joint");
  goal.trajectory.joint_names.push_back("arm_6_joint");
  goal.trajectory.joint_names.push_back("arm_7_joint");

  // Two waypoints in this goal trajectory
  goal.trajectory.points.resize(2); //2

  // First trajectory point
  // Positions
  int index = 0;
  goal.trajectory.points[index].positions.resize(7);
  goal.trajectory.points[index].positions[0] =  0.2;
  goal.trajectory.points[index].positions[1] = 0.0;
  goal.trajectory.points[index].positions[2] =  -1.5;
  goal.trajectory.points[index].positions[3] = 1.94;
  goal.trajectory.points[index].positions[4] = -1.57;
  goal.trajectory.points[index].positions[5] = -0.5;
  goal.trajectory.points[index].positions[6] =  0.0;
  // Velocities
  goal.trajectory.points[index].velocities.resize(7);
  for (int j = 0; j < 7; ++j)
  {
    goal.trajectory.points[index].velocities[j] = 1.0;
  }
  // To be reached 2 second after starting along the trajectory
  goal.trajectory.points[index].time_from_start = ros::Duration(2.0);
  ROS_INFO("done with waypoint ...");
  // Second trajectory point
  // Positions
  index += 1;

  goal.trajectory.points[index].positions.resize(7); //4e-06 -2e-06 0 0 0 -1.62013
  goal.trajectory.points[index].positions[0] = 1.0591656143964283; //0.000004; //2.5; shoulder_pan_joint
  goal.trajectory.points[index].positions[1] = 0.48357964958694666; //-0.000002; //0.2; shoulder_lift_joint
  goal.trajectory.points[index].positions[2] = -1.8060198686571673; //0.00; //-2.1; upper_arm_roll_joint
  goal.trajectory.points[index].positions[3] = 0.28170960172582893; //1.9; elbow_flex_joint
  goal.trajectory.points[index].positions[4] = -1.82175838806662; //0.00; //1.0; forearm_roll_joint
  goal.trajectory.points[index].positions[5] = 0.07656840816165061;  //-1.62013; //-0.5; wrist_flex_joint
  goal.trajectory.points[index].positions[6] = -2.074393506565951; //0.78;  // wrist_roll_joint
  // Velocities
  goal.trajectory.points[index].velocities.resize(7);
  for (int j = 0; j < 7; ++j)
  {
    goal.trajectory.points[index].velocities[j] = 0.0;
  }
  // To be reached 4 seconds after starting along the trajectory
  goal.trajectory.points[index].time_from_start = ros::Duration(4.0);
}


// Entry point
int main(int argc, char** argv)
{
  // Init the ROS node
  ros::init(argc, argv, "run_traj_control");

  ROS_INFO("Starting run_traj_control application ...");
 
  // Precondition: Valid clock
  ros::NodeHandle nh;
  if (!ros::Time::waitForValid(ros::WallDuration(10.0))) // NOTE: Important when using simulated clock
  {
    ROS_FATAL("Timed-out waiting for valid time.");
    return EXIT_FAILURE;
  }

  // Create an arm controller action client to move the TIAGo's arm
  arm_control_client_Ptr ArmClient;
  createArmClient(ArmClient);

  // Generates the goal for the TIAGo's arm
  control_msgs::FollowJointTrajectoryGoal arm_goal;
  waypoints_arm_goal(arm_goal);

  // Sends the command to start the given trajectory 1s from now
  arm_goal.trajectory.header.stamp = ros::Time::now() + ros::Duration(1.0);
  ArmClient->sendGoal(arm_goal);

  // Wait for trajectory execution
  while(!(ArmClient->getState().isDone()) && ros::ok())
  {
    ros::Duration(4).sleep(); // sleep for four seconds
  }

  return EXIT_SUCCESS;
}

File: src/elevator_press_button/scripts-notUsed/elevator_press_button.py
#!/usr/bin/env python
import rospy
#from geometry_msgs.msg import Twist
from trajectory_msgs.msg import JointTrajectory
from control_msgs.msg import FollowJointTrajectoryAction,FollowJointTrajectoryGoal
from std_srvs.srv import Empty
import actionlib
import sys
from copy import copy
from trajectory_msgs.msg import JointTrajectoryPoint #change to tiago


class press_button():

	#def before(self):
		##for each service that we want to use (all the rqt call we do) we can do it  with : 
		##rospy.wait_for_service('global_localization')
		##the wait_for_service func - will make sure the service is up - otherwise we will get notification (i hope so)
		##global_localization = rospy.ServiceProxy('global_localization', Empty)
		##this call trying to connect to the node and sending the message to the server - in our case - empty (but we 			can maybe   use it ##for the arm)'''
		#rospy.init_node('map_navigation', anonymous=False)
		#/arm_controller/follow_joint_trajectory/goal

		    
			
	def __init__(self):
		rospy.init_node('press_button', anonymous=False)
		rospy.loginfo("start arm server")
		#rospy.wait_for_service('/arm_controller/follow_joint_trajectory/goal')
		#arm_control = rospy.ServiceProxy('/arm_controller/follow_joint_trajectory/goal', Empty) 
		#arm_control()

		#ac = actionlib.SimpleActionClient('arm_controller', FollowJointTrajectoryGoal)
      		#wait for the action server to come up
     		#while(not ac.wait_for_server(rospy.Duration.from_sec(5.0))):
              		#rospy.loginfo("Waiting for the move_base action server to come up")


		self._client = actionlib.SimpleActionClient('/arm_controller/follow_joint_trajectory',
		FollowJointTrajectoryAction,
		)
		self._goal = FollowJointTrajectoryGoal()
        	self._goal_time_tolerance = rospy.Time(0.1)
        	self._goal.goal_time_tolerance = self._goal_time_tolerance
        	server_up = self._client.wait_for_server(timeout=rospy.Duration(10.0))
		print("here")
        	if not server_up:
			print("not")
            		rospy.logerr("Timed out waiting for Joint Trajectory")
            		rospy.signal_shutdown("Timed out waiting for Action Server")
            		sys.exit(1)

		#self.clear()	
  
	def points(self):
		rospy.loginfo("sending goal")
       		
		jt = JointTrajectory()
		jt.header.stamp=rospy.Time.now()
		jt.header.frame_id = "/base_link";
		#jt.joint_names = ['arm_1_joint',
                  #             	  'arm_2_joint',
                  #             	  'arm_3_joint',
                   #               'arm_4_joint',
                   #               'arm_5_joint',
                  #                'arm_6_joint',
		#	          'arm_7_joint']

		
		#self._goal.trajectory.append(jt)
		print("after append names")

		point = JointTrajectoryPoint()
		point.positions = [1.0591656143964283, 
		     		   0.48357964958694666, 
		     		   -1.8060198686571673,
	             		   0.28170960172582893, 
		     		   -1.82175838806662,
		     		   0.07656840816165061, 
		     		   -2.074393506565951]

		print("after points")
		point.time_from_start = rospy.Duration(4.0)
		self._goal.trajectory.points.append(point)
		
		#jt.points.append(point)
		print("after append points")
		
		

        def start(self):
                self._goal.trajectory.header.stamp = rospy.Time.now()
	        self._client.send_goal(self._goal)
		print("start is here")
	
        def stop(self):
                self._client.cancel_goal()

      	def wait(self, timeout=15.0):
                self._client.wait_for_result(timeout=rospy.Duration(timeout))

      	#def result(self):
              	#return self._client.get_result()

      
        #def clear(self):
                #self._goal = FollowJointTrajectoryGoal()
                #self._goal.goal_time_tolerance = self._goal_time_tolerance


if __name__ == '__main__':
        try:
                # Testing our function
	        #before()
                #points()
	        #start()
	        arm=press_button()
		arm.points()
		arm.start()
		arm.wait(15.0)
	        rospy.spin()
        except rospy.ROSInterruptException:
                pass

